src/librustc_middle/ty/layout.rs

```diff
@@ -636,8 +636,16 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
             }
 
             ty::Tuple(tys) => {
-                let kind =
-                    if tys.len() == 0 { StructKind::AlwaysSized } else { StructKind::MaybeUnsized };
```

Comment:
This is the important part: `StructKind::MaybeUnsized` prevents the last field from being reordered, and this used to assume that all non-empty tuples were MaybeUnsized

Reply:
Yes and as you discovered, this check is *important* and cannot just be removed.